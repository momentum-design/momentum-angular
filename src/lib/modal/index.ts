export * from './modal.module';
export * from './modal.component';
export * from './modal-body.module';
export * from './modal-body.component';
export * from './modal-footer.module';
export * from './modal-footer.component';
export * from './modal-header.module';
export * from './modal-header.component';

