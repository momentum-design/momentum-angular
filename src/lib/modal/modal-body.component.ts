/** @component modal-body */

import { Component, Input, HostBinding } from '@angular/core';

@Component({
  selector: 'md-modal-body',
  template: `
    <ng-content></ng-content>
  `,
  styles: [],
})
export class ModalBodyComponent {
  /** @prop css class names | '' | '' */
  @Input() public class: string = '';

  @HostBinding('class') get className(): string {
    return 'md-modal__body' + `${(this.class && ` ${this.class}`) || ''}` + ``;
  }
}
