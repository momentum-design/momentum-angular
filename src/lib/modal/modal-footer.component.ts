/** @component modal-footer */

import { Component,  Input, HostBinding } from '@angular/core';

@Component({
  selector: 'md-modal-footer',
  template: `
    <ng-content></ng-content>
  `,
  styles: [],
})
export class ModalFooterComponent  {
  /** @prop css class names | '' | '' */
  @Input() public class: string = '';

  @HostBinding('class') get className(): string {
    return (
      'md-modal__footer' + `${(this.class && ` ${this.class}`) || ''}` + ``
    );
  }
}
