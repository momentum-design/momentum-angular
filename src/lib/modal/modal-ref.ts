import { OverlayRef } from '@angular/cdk/overlay';
import { Subject } from 'rxjs';
import { TemplateRef, Type } from '@angular/core';

interface ModalCloseEvent {
   type: 'backdropClick' | 'close';
   data: object;
}

export type ModalContent = TemplateRef<any> | Type<any>;

export class ModalRef {
  private onHide = new Subject<ModalCloseEvent>();
  onHide$ = this.onHide.asObservable();

  constructor(
    public overlay: OverlayRef,
    public content: ModalContent,
    public data: object,
    public backdropClickExit: boolean) {
      if (backdropClickExit) {
        overlay.backdropClick().subscribe(() => {
          this._close('backdropClick', null); });
      }
  }

  close(data?: object) {
    this._close('close', data);
  }

  private _close(type: ModalCloseEvent['type'], data?: object) {
    this.overlay.detach();
    this.onHide.next({
      type,
      data
    });
    this.onHide.complete();
  }
}

