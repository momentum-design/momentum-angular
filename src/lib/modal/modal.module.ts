import { NgModule, ModuleWithProviders, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { CommonModule } from '@angular/common';
import { A11yModule } from '@angular/cdk/a11y';
import { OverlayModule } from '@angular/cdk/overlay';
import { ModalComponent } from './modal.component';
import { ModalBodyComponent } from './modal-body.component';
import { ModalFooterComponent } from './modal-footer.component';
import { ModalHeaderComponent } from './modal-header.component';
import { ModalContainerComponent } from './modal-container.component';
import { ModalService } from './modal.service';



@NgModule({
  declarations: [
    ModalComponent,
    ModalBodyComponent,
    ModalFooterComponent,
    ModalHeaderComponent,
    ModalContainerComponent
  ],
  imports: [
    CommonModule,
    A11yModule,
    OverlayModule
  ],
  exports: [
    ModalComponent,
    ModalBodyComponent,
    ModalFooterComponent,
    ModalHeaderComponent,
    ModalContainerComponent
  ],
  entryComponents: [ModalContainerComponent],
  providers: [ModalService],
  schemas: [CUSTOM_ELEMENTS_SCHEMA]
})
export class ModalModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: ModalModule,
      providers: [],
    };
  }
}
