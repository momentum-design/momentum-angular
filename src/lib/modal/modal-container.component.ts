/** @component modal-container */

import {
  Component,
  OnInit
} from '@angular/core';
import { ModalContent, ModalRef } from './modal-ref';

@Component({
  selector: 'md-modal-container',
  template: `
  <ng-container>
    <ng-container *ngComponentOutlet="content"></ng-container>
  </ng-container>
  `,
})
export class ModalContainerComponent  implements OnInit {

   public content: ModalContent;
   public context;

   constructor( private modalRef: ModalRef) {}

  ngOnInit() {
    this.content = this.modalRef.content;
    this.context = {
      close: this.modalRef.close.bind(this.modalRef)
    };

  }

}
