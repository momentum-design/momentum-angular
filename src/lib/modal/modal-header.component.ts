/** @component modal-header */

import {
  Component,
  Input,
  HostBinding,
  HostListener,
} from '@angular/core';
import { ModalRef } from './modal-ref';


@Component({
  selector: 'md-modal-header',
  template: `
  <span class="md-modal__title">{{headerLabel}}</span>
  <span *ngIf="message" class="md-modal__message">{{
    message
  }}</span>
  <button
    *ngIf="showCloseButton"
    class="md-close md-modal__close"
    (click)="close()"
    (keydown)="handleKeydown($event)"
    aria-label="close"
  >
  </button>
  `,
})
export class ModalHeaderComponent {
  /** @prop Optional css class string | '' */
  @Input() public class: string = '';
  /** @prop ModalHeader label text | '' */
  @Input() public headerLabel: string = '';
  /** @prop  Modal message | '' */
  @Input() public message: string = '';
  /** @prop show/hide close button | true */
  @Input() public showCloseButton: boolean = true;

  @HostBinding('class') get className(): string {
    return (
      'md-modal__header' + `${(this.class && ` ${this.class}`) || ''}` + ``
    );
  }


  @HostListener('document:keydown.escape', ['$event']) handleKeydown(
    _: KeyboardEvent
  ) {
    this.close();
  }

  constructor(private modalRef: ModalRef) {}

  close() {
    this.modalRef.close();
  }
}
