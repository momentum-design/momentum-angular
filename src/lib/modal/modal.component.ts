/** @component modal */

import {
  Component,
  Input,
  OnInit,
  ElementRef,
  AfterViewInit,
} from '@angular/core';
import { FocusTrapFactory } from '@angular/cdk/a11y';
import { ModalContent, ModalRef } from './modal-ref';

export type SizeType =
  | 'large'
  | 'medium'
  | 'default'
  | 'small'
  | 'full'
  | 'dialog';

@Component({
  selector: 'md-modal',
  template: `
    <div
      role="dialog"
      id="{{htmlId}}"
      class="md-modal md-modal--{{sizeType}} in {{class}}"
      attr.aria-labelledby="{{ariaLabel}}"
      aria-modal="true"
    >
      <div class="md-modal__content" cdkTrapFocusAutoCapture >
        <div class="md-modal__flex-container">
          <ng-content></ng-content>
        </div>
      </div>
    </div>
  `,
  styles: [`
  .md-modal.md-modal--full {
      position: static;
    }
  `]
})
export class ModalComponent implements OnInit, AfterViewInit {
  /** @prop Determines the visibility and ability to edit the backdrop of the Modal | true */
  @Input() public backdrop: Boolean = true;

  private _classList: '';
  /** @propptional css class names | '' */
  @Input()
  set class(value) {
    this._classList = value;
  }
  get class() {
    return this._classList;
  }

  /** @prop size of the modal | 'default' */
  @Input() public sizeType: SizeType = 'default';

  /** @prop htmlId for modal | 'md-modal' */
  @Input() public htmlId: String = 'md-modal';

   /** @prop ariaLabel for modal | '' */
   @Input() public ariaLabel: String = '';

   /** @option data array of data for modal | [] */
   @Input() public data: [];


   public content: ModalContent;
   public context;

   constructor(
      private modalRef: ModalRef,
      private _elRef: ElementRef,
      private _focusTrapFactory: FocusTrapFactory
    ) {}

    ngAfterViewInit() {
      this._trapFocus();
    }

    ngOnInit() {
      this.content = this.modalRef.content;
      this.context = {
        close: this.modalRef.close.bind(this.modalRef)
      };

    }

    private _trapFocus() {
      // creates a focus trap region
      const focusTrap = this._focusTrapFactory.create(this._elRef.nativeElement);
      // Moves the focus in the form (by default the first field)
      focusTrap.focusInitialElement();
    }

}
