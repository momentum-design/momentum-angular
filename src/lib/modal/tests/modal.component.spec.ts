import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { ModalService } from '../modal.service';
import { Overlay } from '@angular/cdk/overlay';
import { ModalComponent } from '../modal.component';

describe('ModalComponent', () => {
  let component: ModalComponent;
  let fixture: ComponentFixture<ModalComponent>;
  let testElement: HTMLElement;
  let testService: Overlay;
  let textModalService: ModalService;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [ModalComponent],
      providers: [ModalService, Overlay],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(ModalComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create and match snapshot', () => {
    expect(component).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });

  it('should render full fix', () => {
    component.ngOnInit();
    component.showModal();
    fixture.detectChanges();
    testElement = fixture.nativeElement;
    expect(testElement.innerHTML).toBeTruthy();
  });

  it('should run service Overlay', () => {
    component.ngOnInit();
    fixture.detectChanges();
    testElement = fixture.nativeElement;
    testService = TestBed.get(Overlay);
    expect(testService).toBeTruthy();
  });

  it('should run textModalService Overlay', () => {
    component.ngOnInit();
    fixture.detectChanges();
    testElement = fixture.nativeElement;
    textModalService = TestBed.get(Overlay);
    expect(textModalService).toBeTruthy();
  });
});
