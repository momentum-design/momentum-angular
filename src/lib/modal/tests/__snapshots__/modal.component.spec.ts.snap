// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalComponent should create and match snapshot 1`] = `
<md-modal
  backdrop={[Function Boolean]}
  backdropClickExit="false"
  closeModal={[Function Function]}
  dismissContent={[Function Function]}
  modalContentTemplate={[Function TemplateRef_]}
  modalService={[Function ModalService]}
  overlay={[Function Overlay]}
  overlayRef={[Function OverlayRef]}
  show="false"
  showContent={[Function Function]}
  showModal={[Function Function]}
  sizeType={[Function String]}
  tp={[Function TemplatePortal]}
  viewContainerRef={[Function ViewContainerRef_]}
  whenHide={[Function EventEmitter]}
>
  
</md-modal>
`;
